byzcoin/struct.go

```diff
@@ -194,77 +120,131 @@ func (s *stateChangeStorage) setMaxNbrBlock(nbr int) {
 	s.maxNbrBlock = nbr
 }
 
+// calculateSize reads the entries in the database and sums up their
+// sizes
+func (s *stateChangeStorage) calculateSize() error {
+	s.Lock()
+	defer s.Unlock()
+
+	s.size = 0
+
+	return s.db.View(func(tx *bbolt.Tx) error {
+		b := tx.Bucket(s.bucket)
+		if b == nil {
+			return errors.New("Missing bucket")
+		}
+
+		return b.ForEach(func(scid, v []byte) error {
+			scb := b.Bucket(scid)
+			if scb == nil {
+				return nil
+			}
+
+			return scb.ForEach(func(k, v []byte) error {
+				s.size += len(v)
+				return nil
+			})
+		})
+	})
+}
+
 // This will clean the oldest state changes when the total size
 // is above the maximum. It will remove elements until cleanThreshold of
 // the space is available.
 func (s *stateChangeStorage) cleanBySize() error {
-	// size and sortedKeys need to be concurrent safe
-	s.sortedKeysLock.Lock()
-	defer s.sortedKeysLock.Unlock()
-
 	if s.size < s.maxSize || s.maxSize == 0 {
 		// nothing to clean
 		return nil
 	}
 
-	sortedKeys := make(keyTimeArray, len(s.sortedKeys))
-	copy(sortedKeys, s.sortedKeys)
+	size := s.size
```

Comment:
do you need a mutex lock here?

Reply:
The lock is already held by the append function